File: mac_scripts/functions/brew-desired-update-js/manage.mjs
import { brewInfoToken, createBrewRunner, resolveBrewBinary } from "./brew.mjs";
import {
  loadDesiredDocument,
  normalizeDesiredDocument,
  resolveBudPaths,
  writeDesiredDocument,
} from "./config.mjs";

/**
 * @param {string} name
 */
export function isTapFormulaSpec(name) {
  return name.split("/").length === 3;
}

/**
 * @param {string} name
 */
export function tapFromFormulaSpec(name) {
  return name.slice(0, name.lastIndexOf("/"));
}

/**
 * @param {string} name
 */
export function formulaNameFromSpec(name) {
  return name.slice(name.lastIndexOf("/") + 1);
}

/**
 * @param {string} token
 * @param {string} name
 */
function tokenMatchesName(token, name) {
  return token === name || token.slice(token.lastIndexOf("/") + 1) === name;
}

/**
 * @param {string[]} argv
 */
export function parseManageArgs(argv) {
  let forceType = null;
  const names = [];
  for (const arg of argv) {
    if (arg === "--cask") {
      forceType = "cask";
    } else if (arg === "--formula") {
      forceType = "formula";
    } else if (arg === "--tap") {
      forceType = "tap";
    } else {
      names.push(arg);
    }
  }
  return { forceType, names };
}

/**
 * @param {string} name
 * @param {object} deps
 */
export async function detectBrewType(name, deps) {
  const {
    brewBin,
    runBrew: runBrewFn,
    ui,
    brewInfoToken: infoToken = brewInfoToken,
  } = deps;

  const formulaToken = await infoToken("formula", name, { brewBin, runBrew: runBrewFn });
  const caskToken = await infoToken("cask", name, { brewBin, runBrew: runBrewFn });

  const isFormula = Boolean(formulaToken && tokenMatchesName(formulaToken, name));
  const isCask = Boolean(caskToken && tokenMatchesName(caskToken, name));

  if (isCask && isFormula) {
    ui?.warn(`'${name}' exists as both cask and formula`);
    ui?.info(`Use 'budj add --cask ${name}' or 'budj add --formula ${name}'`);
    return { error: "dual" };
  }
  if (isCask) {
    return { type: "cask" };
  }
  if (isFormula) {
    return { type: "formula" };
  }

  ui?.warn(`'${name}' not found in Homebrew`);
  ui?.info(`Use 'brew search ${name}' to find similar packages`);
  return { error: "missing" };
}

/**
 * @param {string[]} list
 * @param {string} item
 */
function addUnique(list, item) {
  if (list.includes(item)) {
    return false;
  }
  list.push(item);
  return true;
}

/**
 * @param {string} pkgName
 * @param {object} options
 */
async function applyOneAdd(pkgName, { forceType, document, deps, detect }) {
  const { ui, brewBin, runBrew: runBrewFn } = deps;
  const runner = runBrewFn ?? (brewBin ? createBrewRunner({ brewBin, ui }) : null);
  const brewDeps = runner ? { ...deps, runBrew: runner } : deps;

  let pkgType = forceType;
  let listName = pkgName;
  let tapForFormula = "";
  let detectedViaBrew = false;
  let validatedViaBrew = false;

  if (!pkgType) {
    if (isTapFormulaSpec(pkgName)) {
      pkgType = "formula";
      listName = formulaNameFromSpec(pkgName);
      tapForFormula = tapFromFormulaSpec(pkgName);
    } else if (pkgName.includes("/")) {
      pkgType = "tap";
    } else {
      const detected = await detect(pkgName, brewDeps);
      if (detected.error) {
        return false;
      }
      pkgType = detected.type;
      detectedViaBrew = true;
    }
  } else if (pkgType === "tap") {
    if (!pkgName.includes("/") || isTapFormulaSpec(pkgName)) {
      ui?.warn("Tap name must be user/repo (e.g. mongodb/brew)");
      return false;
    }
  } else if (pkgType === "formula" && isTapFormulaSpec(pkgName)) {
    listName = formulaNameFromSpec(pkgName);
    tapForFormula = tapFromFormulaSpec(pkgName);
  } else if ((pkgType === "cask" || pkgType === "formula") && runner) {
    const flag = pkgType === "cask" ? "--cask" : "--formula";
    const info = await runner(["info", flag, pkgName]);
    if (info.code !== 0) {
      ui?.warn(`'${pkgName}' not found as a Homebrew ${pkgType}`);
      return false;
    }
    validatedViaBrew = true;
  }

  const listKey = pkgType === "cask" ? "casks" : pkgType === "formula" ? "formulas" : "taps";
  const label = pkgType === "cask" ? "cask" : pkgType === "formula" ? "formula" : "tap";

  if (document[listKey].includes(listName)) {
    ui?.warn(`'${listName}' already exists in ${label}s list`);
    return false;
  }

  if (pkgType === "formula" && !tapForFormula && runner) {
    const infoToken = deps.brewInfoToken ?? brewInfoToken;
    const token = await infoToken("formula", pkgName, { brewBin, runBrew: runner });
    if (token?.includes("/")) {
      tapForFormula = token.slice(0, token.lastIndexOf("/"));
    }
  }

  if (pkgType === "formula" && !detectedViaBrew && !validatedViaBrew && !isTapFormulaSpec(pkgName) && runner) {
    const info = await runner(["info", "--formula", pkgName]);
    if (info.code !== 0) {
      ui?.warn(`'${pkgName}' not found as a Homebrew formula`);
      return false;
    }
  }

  if (tapForFormula) {
    if (addUnique(document.taps, tapForFormula)) {
      ui?.info(`Added '${tapForFormula}' to taps (now ${document.taps.length})`);
    }
  }

  document[listKey].push(listName);
  ui?.info(`Added '${listName}' to ${label}s (now ${document[listKey].length})`);
  return true;
}

/**
 * @param {string[]} names
 * @param {{ forceType: string | null, document: object, deps: object }} options
 */
export async function applyAdd(names, { forceType, document, deps }) {
  const doc = {
    version: document.version,
    formulas: [...document.formulas],
    casks: [...document.casks],
    taps: [...document.taps],
  };
  const detect = deps.detectBrewType ?? detectBrewType;
  let succeeded = 0;
  let failed = 0;

  for (const name of names) {
    const ok = await applyOneAdd(name, { forceType, document: doc, deps, detect });
    if (ok) {
      succeeded += 1;
    } else {
      failed += 1;
    }
  }

  return {
    document: normalizeDesiredDocument(doc),
    succeeded,
    failed,
  };
}

/**
 * @param {string[]} names
 * @param {{ document: object, ui?: object }} options
 */
export function applyRemove(names, { document, ui }) {
  const doc = {
    version: document.version,
    formulas: [...document.formulas],
    casks: [...document.casks],
    taps: [...document.taps],
  };
  let succeeded = 0;
  let failed = 0;

  for (const name of names) {
    if (doc.casks.includes(name)) {
      doc.casks = doc.casks.filter((item) => item !== name);
      ui?.info(`Removed '${name}' from casks (now ${doc.casks.length})`);
      succeeded += 1;
    } else if (doc.formulas.includes(name)) {
      doc.formulas = doc.formulas.filter((item) => item !== name);
      ui?.info(`Removed '${name}' from formulas (now ${doc.formulas.length})`);
      succeeded += 1;
    } else if (doc.taps.includes(name)) {
      doc.taps = doc.taps.filter((item) => item !== name);
      ui?.info(`Removed '${name}' from taps (now ${doc.taps.length})`);
      succeeded += 1;
    } else {
      ui?.warn(`'${name}' not found in casks, formulae, or taps list`);
      failed += 1;
    }
  }

  return {
    document: normalizeDesiredDocument(doc),
    succeeded,
    failed,
  };
}

/**
 * @param {string[]} args
 * @param {object} context
 */
export async function runAddCommand(args, context = {}) {
  const {
    env = process.env,
    ui,
    loadDesired = loadDesiredDocument,
    resolveBrew = resolveBrewBinary,
    writeDesired = writeDesiredDocument,
    add = applyAdd,
  } = context;

  const { forceType, names } = parseManageArgs(args);
  if (names.length === 0) {
    ui.error("Package name is required for add action");
    return 1;
  }

  const brewBin = resolveBrew({ env });
  if (!brewBin) {
    ui.error("Homebrew not found.");
    return 1;
  }

  const loaded = await loadDesired({ env });
  if (!loaded.ok) {
    ui.error(loaded.error);
    return 1;
  }

  const { document, succeeded, failed } = await add(names, {
    forceType,
    document: loaded.document,
    deps: { ui, brewBin, runBrew: context.runBrew ?? createBrewRunner({ brewBin, ui }) },
  });

  if (succeeded > 0) {
    const { desiredFile } = resolveBudPaths(env);
    const written = writeDesired(desiredFile, document);
    if (!written.ok) {
      ui.error(written.error);
      return 1;
    }
    ui.info(`Changes saved to ${desiredFile}`);
  }

  return failed > 0 ? 1 : 0;
}

/**
 * @param {string[]} args
 * @param {object} context
 */
export async function runRemoveCommand(args, context = {}) {
  const {
    env = process.env,
    ui,
    loadDesired = loadDesiredDocument,
    writeDesired = writeDesiredDocument,
    remove = applyRemove,
  } = context;

  const { names } = parseManageArgs(args);
  if (names.length === 0) {
    ui.error("Package name is required for remove action");
    return 1;
  }

  const loaded = await loadDesired({ env });
  if (!loaded.ok) {
    ui.error(loaded.error);
    return 1;
  }

  const { document, succeeded, failed } = remove(names, {
    document: loaded.document,
    ui,
  });

  if (succeeded > 0) {
    const { desiredFile } = resolveBudPaths(env);
    const written = writeDesired(desiredFile, document);
    if (!written.ok) {
      ui.error(written.error);
      return 1;
    }
    ui.info(`Changes saved to ${desiredFile}`);
  }

  return failed > 0 ? 1 : 0;
}

File: mac_scripts/functions/brew-desired-update/manage.py
from brew import ensure_tap
from config import normalize
from detect import detect_brew_type, info_token, parse_info_token
from messages import fmt

SKIP_CASK_TAPS = frozenset({"homebrew/cask", "homebrew/cask-fonts"})
LIST_KEY = {"cask": "casks", "formula": "formulas", "tap": "taps"}


def _is_tap_formula_spec(name: str) -> bool:
    return len(name.split("/")) == 3


def _tap_from_formula_spec(name: str) -> str:
    return name[: name.rfind("/")]


def _formula_name_from_spec(name: str) -> str:
    return name[name.rfind("/") + 1 :]


def _copy_doc(document: dict) -> dict:
    return {
        "version": document["version"],
        "formulas": list(document["formulas"]),
        "casks": list(document["casks"]),
        "taps": list(document["taps"]),
    }


def _assign_doc(document: dict, updated: dict) -> None:
    document["version"] = updated["version"]
    document["formulas"] = updated["formulas"]
    document["casks"] = updated["casks"]
    document["taps"] = updated["taps"]


def _cask_tap_from_search(stdout: str, pkg_name: str) -> str | None:
    for line in stdout.splitlines():
        if pkg_name in line and "/" in line:
            parts = line.strip().split("/")
            if len(parts) < 2:
                return None
            tap_name = f"{parts[0]}/{parts[1]}"
            if tap_name in SKIP_CASK_TAPS:
                return None
            return tap_name
    return None


def _listed_taps(runner, ui) -> list[str]:
    result = runner.run(["tap"], ui)
    return [line for line in (result.stdout or "").splitlines() if line]


def _ensure_cask_tap(pkg_name: str, runner, ui) -> bool:
    search = runner.run(["search", "--cask", pkg_name], ui)
    tap_name = _cask_tap_from_search(search.stdout or "", pkg_name)
    if not tap_name:
        return True
    if tap_name in _listed_taps(runner, ui):
        return True
    ui.step(f"Adding required tap: {tap_name}")
    if not ensure_tap(tap_name, runner):
        ui.error(fmt("tap.failed", name=tap_name))
        return False
    return True


def _add_one(pkg_name: str, force_type: str | None, doc: dict, runner, ui) -> bool:
    ui.step(f"Adding {pkg_name}")
    pkg_type = force_type
    list_name = pkg_name
    tap_for_formula = ""
    info_stdout = ""

    if not pkg_type:
        if _is_tap_formula_spec(pkg_name):
            pkg_type = "formula"
            list_name = _formula_name_from_spec(pkg_name)
            tap_for_formula = _tap_from_formula_spec(pkg_name)
        elif "/" in pkg_name:
            pkg_type = "tap"
        else:
            pkg_type = detect_brew_type(pkg_name, runner, ui)
            if pkg_type is None:
                return False
    elif pkg_type == "tap":
        if "/" not in pkg_name or _is_tap_formula_spec(pkg_name):
            ui.warn(fmt("tap.name.invalid"))
            return False
    elif pkg_type == "formula" and _is_tap_formula_spec(pkg_name):
        list_name = _formula_name_from_spec(pkg_name)
        tap_for_formula = _tap_from_formula_spec(pkg_name)
    elif pkg_type in ("cask", "formula"):
        result = runner.run(["info", f"--{pkg_type}", pkg_name], ui)
        if result.returncode != 0:
            ui.warn(fmt("not.found.type", name=pkg_name, type=pkg_type))
            return False
        info_stdout = result.stdout or ""

    if pkg_type == "formula" and not tap_for_formula:
        token = (
            parse_info_token(info_stdout)
            if info_stdout
            else info_token("formula", pkg_name, runner, ui)
        )
        if token and "/" in token:
            tap_for_formula = token[: token.rfind("/")]

    list_key = LIST_KEY[pkg_type]
    if list_name in doc[list_key]:
        ui.warn(fmt("already.in.list", name=list_name, type=pkg_type))
        return False

    if tap_for_formula and tap_for_formula not in doc["taps"]:
        doc["taps"].append(tap_for_formula)
        ui.active(fmt("added.tap", name=tap_for_formula, n=len(doc["taps"])))

    if pkg_type == "cask" and not _ensure_cask_tap(pkg_name, runner, ui):
        return False

    doc[list_key].append(list_name)
    ui.active(
        fmt("added.item", name=list_name, type=pkg_type, n=len(doc[list_key]))
    )
    return True


def run_add(names, force_type, document, runner, ui, save_fn) -> int:
    doc = _copy_doc(document)
    succeeded = 0
    failed = 0
    for name in names:
        if _add_one(name, force_type, doc, runner, ui):
            succeeded += 1
        else:
            failed += 1
    if succeeded:
        _assign_doc(document, normalize(doc))
        save_fn(document)
    return 1 if failed else 0


def run_remove(names, document, ui, save_fn) -> int:
    doc = _copy_doc(document)
    succeeded = 0
    failed = 0
    for name in names:
        ui.step(f"Removing {name}")
        if name in doc["casks"]:
            doc["casks"] = [item for item in doc["casks"] if item != name]
            ui.active(fmt("removed.item", name=name, type="cask", n=len(doc["casks"])))
            succeeded += 1
        elif name in doc["formulas"]:
            doc["formulas"] = [item for item in doc["formulas"] if item != name]
            ui.active(
                fmt("removed.item", name=name, type="formula", n=len(doc["formulas"]))
            )
            succeeded += 1
        elif name in doc["taps"]:
            doc["taps"] = [item for item in doc["taps"] if item != name]
            ui.active(fmt("removed.item", name=name, type="tap", n=len(doc["taps"])))
            succeeded += 1
        else:
            ui.warn(fmt("not.in.list", name=name))
            failed += 1
    if succeeded:
        _assign_doc(document, normalize(doc))
        save_fn(document)
    return 1 if failed else 0
